pf/internal/configencoding/provider.go

```diff
@@ -0,0 +1,120 @@
+
+	pl "github.com/pulumi/pulumi-terraform-bridge/pf/internal/plugin"
+	"github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfbridge"
+)
+
+type ProviderWithContextAndConfig interface {
+	pl.ProviderWithContext
+
+	Attach(address string) error
+
+	// GetConfigEncoding supplies the wrapping [Provider] with the (possibly unstable)
+	// encoding used by the wrapped ProviderWithContextAndConfig.
+	//
+	// This is necessary since dynamic providers change their encoding in response to
+	// gRPC calls.
+	GetConfigEncoding(context.Context) *tfbridge.ConfigEncoding
+}
+
+type provider[T any] struct{ ProviderWithContextAndConfig }
+
+type Provider[T any] interface {
+	pl.ProviderWithContext
+
+	Unwrap() T
+}
+
+// New creates a new [plugin.Provider] that wraps prov.
+//
+// The returned provider will correctly handle JSON encoded property values according to
+// the passed encoding.
+func New[T ProviderWithContextAndConfig](prov T) Provider[T] {
+	return &provider[T]{prov}
+}
+
+// Unwrap returns the original [pl.ProviderWithContext] used to create the [Provider].
+func (p *provider[T]) Unwrap() T { return p.ProviderWithContextAndConfig.(T) }
+
+func (p *provider[T]) CheckConfigWithContext(
+	ctx context.Context, urn resource.URN, olds, news resource.PropertyMap,
+	allowUnknowns bool,
+) (resource.PropertyMap, []plugin.CheckFailure, error) {
+	encoding := p.GetConfigEncoding(ctx)
+	olds, err := encoding.UnfoldProperties(olds)
+	if err != nil {
+		return nil, nil, err
+	}
+	news, err = encoding.UnfoldProperties(news)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	props, failures, err := p.ProviderWithContextAndConfig.
+		CheckConfigWithContext(ctx, urn, olds, news, allowUnknowns)
+	if err != nil {
+		return nil, nil, err
+	}
+	props, err = encoding.FoldProperties(props)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return props, failures, err
+}
+
+func (p *provider[T]) DiffConfigWithContext(
+	ctx context.Context, urn resource.URN, oldInputs, olds, news resource.PropertyMap,
+	allowUnknowns bool, ignoreChanges []string) (plugin.DiffResult, error) {
+	encoding := p.GetConfigEncoding(ctx)
+	oldInputs, err := encoding.UnfoldProperties(oldInputs)
+	if err != nil {
+		return plugin.DiffConfigResponse{}, err
+	}
+
+	olds, err = encoding.UnfoldProperties(olds)
+	if err != nil {
+		return plugin.DiffConfigResponse{}, err
+	}
+
+	news, err = encoding.UnfoldProperties(news)
+	if err != nil {
+		return plugin.DiffConfigResponse{}, err
+	}
+
+	return p.ProviderWithContextAndConfig.DiffConfigWithContext(
+		ctx, urn, oldInputs, olds, news, allowUnknowns, ignoreChanges)
+}
+
+func (p *provider[T]) ConfigureWithContext(ctx context.Context, inputs resource.PropertyMap) error {
+	inputs, err := p.GetConfigEncoding(ctx).UnfoldProperties(inputs)
+	if err != nil {
+		return err
+	}
+
+	return p.ProviderWithContextAndConfig.ConfigureWithContext(ctx, inputs)
+}
+
+func (p *provider[T]) Attach(address string) error {
```

Comment:
Is attach overload needed?

Reply:
It is not (and I removed it). The [`Attach` method](https://github.com/pulumi/pulumi-terraform-bridge/blob/63e2e72ebf1dba6cfa7cdec645ead9d21867009f/pf/internal/configencoding/provider.go#L30) on `ProviderWithContextAndConfig` is needed however. Go type casting is weird.

Explanation: https://go.dev/play/p/nfWG2seoG0s